### PR TITLE
Gracefully shut down spdk_tgt

### PR DESCRIPTION
**Pull request by derekbit**
Longhorn/longhorn#7263

deps on https://github.com/longhorn/longhorn-spdk-engine/pull/68

### CODE REVIEW

File: app/cmd/start.go [hunk 1]
```diff
@@ -146,89 +154,176 @@ func start(c *cli.Context) (err error) {
 		return err
 	}
 
+	var wg sync.WaitGroup
+
+	waitCount := 4
+	if spdkEnabled {
+		waitCount++
+	}
+	wg.Add(waitCount)
```
Comment (innobead) [inline app/cmd/start.go:163]: Suggest removing this, but execute `wg.Add(1)` when needed later.

File: app/cmd/start.go [hunk 2]
```diff
@@ -146,89 +154,176 @@ func start(c *cli.Context) (err error) {
...
+
 	// Start disk server
-	diskGRPCServer, diskGRPCListener, err := setupDiskGRPCServer(diskServiceAddress, spdkServiceAddress, spdkEnabled, shutdownCh)
+	diskGRPCServer, diskGRPCListener, err := setupDiskGRPCServer(diskServiceAddress, spdkServiceAddress, spdkEnabled, shutdown.ch)
 	if err != nil {
 		return err
 	}
 	go func() {
+		defer func() {
```
Comment (innobead) [inline app/cmd/start.go:171]: Add `wg.Add(1)` just before the `go func() {`. The concept here is to request when needed, so we don't need to pre-count the number we need for the wait group.

File: app/cmd/start.go [hunk 3]
```diff
@@ -146,89 +154,176 @@ func start(c *cli.Context) (err error) {
...
 		pmGRPCServer.Stop()
+		proxyGRPCServer.Stop()
+		diskGRPCServer.Stop()
 	}()
 
-	return <-shutdownCh
+	wg.Wait()
+
+	if spdkEnabled {
+		stopSPDKTgt(30 * time.Second)
```
Comment (innobead) [inline app/cmd/start.go:291]: We start `spdk_tgt` in longhorn-spdk-engine, so can we do stop there as well to make responsibility segregation? The lifecycle of it should be together with the `spdkGRPCServer`. Is this feasible?

File: pkg/util/util.go [hunk 4]
```diff
@@ -135,3 +136,43 @@ func IsSPDKTgtReady(timeout time.Duration) bool {
 	}
 	return false
 }
+
+// FindProcessByCmdline finds the process with matching cmdline
+func FindProcessByCmdline(cmdline string) (*os.Process, error) {
```
Comment (innobead) [inline pkg/util/util.go:141]: Does this worth moving to common-lib?

**Review by innobead — CHANGES_REQUESTED** [review]
The concept LG, but having some comments needed to resolve first.

File: pkg/util/util.go [hunk 5]
```diff
@@ -135,3 +136,43 @@ func IsSPDKTgtReady(timeout time.Duration) bool {
...
+	defer procDir.Close()
+
+	processes, err := procDir.Readdir(-1)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, process := range processes {
+		if process.IsDir() {
+			pid, convErr := strconv.Atoi(process.Name())
```
Comment (c3y1huang) [inline pkg/util/util.go:155]: Could we utilize [GetProcessPID](https://github.com/longhorn/go-common-libs/blob/main/proc/proc.go#L58)?

File: app/cmd/start.go [hunk 6]
```diff
@@ -146,89 +154,176 @@ func start(c *cli.Context) (err error) {
...
+	}
+
+	logrus.Infof("Stopping spdk_tgt process")
+	if err := process.Signal(syscall.SIGTERM); err != nil {
+		logrus.WithError(err).Errorf("Failed to send SIGTERM to spdk_tgt process")
+	} else {
+		done := make(chan error, 1)
+		go func() {
+			_, err := process.Wait()
+			done <- err
```
Comment (c3y1huang) [inline app/cmd/start.go:313]: Should we close the done channel at some point?

File: app/cmd/start.go [hunk 7]
```diff
@@ -146,89 +154,176 @@ func start(c *cli.Context) (err error) {
...
+		}()
+
+		select {
+		case <-time.After(timeout):
+			logrus.Error("spdk_tgt process did not exit in time")
```
Comment (c3y1huang) [inline app/cmd/start.go:318]: I am thinking if we should have an additional SIGKILL to forcibly kill the process if it doesn't terminate within the timeout. WDYT?

Comment (derekbit) [reply app/cmd/start.go:318]: Using `SIGKILL` sounds good.

File: app/cmd/start.go (hunk 3 above)
Comment (derekbit) [reply app/cmd/start.go:291]: We start spdk_tgt before starting longhorn-instance-manager ([code](https://github.com/longhorn/longhorn-instance-manager/blob/master/package/instance-manager#L63)). longhorn-spdk-engine ensures the health of spdk_tgt and restarts it if it is somehow crashed.

Therefore, stopping spdk_tgt before exiting longhorn-instance-manager looks reasonable. WDYT?

File: pkg/util/util.go (hunk 4 above)
Comment (derekbit) [reply pkg/util/util.go:141]: Yes.

File: app/cmd/start.go (hunk 6 above)
Comment (derekbit) [reply app/cmd/start.go:313]: Good catch. Thank you.

File: app/cmd/start.go (hunk 1 above)
Comment (derekbit) [reply app/cmd/start.go]: Remove WaitGroup. Use errgroup instead.

File: pkg/util/util.go (hunk 4 above)
Comment (derekbit) [reply pkg/util/util.go]: Addressed. Thank you.

File: app/cmd/start.go (hunk 7 above)
Comment (derekbit) [reply app/cmd/start.go]: Added. Thank you.

File: app/cmd/start.go (hunk 2 above)
Comment (derekbit) [reply app/cmd/start.go]: Remove WaitGroup. Use errgroup instead.

File: app/cmd/start.go (hunk 3 above)
Comment (innobead) [reply app/cmd/start.go]: Sounds good.

**Review by c3y1huang — APPROVED** [review]
lgtm

**Comment by derekbit** [comment]
Updated the deps. cc @innobead

**Review by innobead — APPROVED** [review]
LGTM

**Comment by innobead** [comment]
@derekbit it's good to have refactoring by replacing the shutdown channel with context to make implementations consistent.

**Pull request merged**